tools/compiler/compiler.go

```diff
@@ -0,0 +1,515 @@
+		for _, p := range o.GetParams() {
+			t, err := types.ProtoAsType(p)
+			if err != nil {
+				return nil, err
+			}
+			args = append(args, env.SerializeTypeDesc(t))
+		}
+		res, err := types.ProtoAsType(o.GetResultType())
+		if err != nil {
+			return nil, err
+		}
+		ret := env.SerializeTypeDesc(res)
+		if o.IsInstanceFunction {
+			overloads = append(overloads, env.NewMemberOverload(o.GetOverloadId(), args[0], args[1:], ret))
+		} else {
+			overloads = append(overloads, env.NewOverload(o.GetOverloadId(), args, ret))
+		}
+	}
+	return env.NewFunction(decl.GetName(), overloads...), nil
+}
+
+// TypeDescriptorSetFile returns an EnvOption which loads type descriptors from a file.
+// The file must be in binary format.
+func TypeDescriptorSetFile(path string) cel.EnvOption {
+	return func(e *cel.Env) (*cel.Env, error) {
+		format := inferFileFormat(path)
+		if format != BinaryProto {
+			return nil, fmt.Errorf("type descriptor must be in binary format")
+		}
+		fds := &descpb.FileDescriptorSet{}
+		if err := loadProtoFile(path, BinaryProto, fds); err != nil {
+			return nil, err
+		}
+		var err error
+		e, err = e.Extend(cel.TypeDescs(fds))
+		if err != nil {
+			return nil, fmt.Errorf("e.Extend() with type descriptor set %v failed: %w", fds, err)
+		}
+		return e, nil
+	}
+}
+
+// InputExpression is an interface for an expression which can be compiled into a CEL AST and return
+// an optional policy metadata map.
+type InputExpression interface {
+	// CreateAst creates a CEL AST from the input expression using the provided compiler.
+	CreateAst(Compiler) (*cel.Ast, map[string]any, error)
+}
+
+// CompiledExpression is an InputExpression which loads a CheckedExpr from a file.
+type CompiledExpression struct {
+	Path string
+}
+
+// CreateAst creates a CEL AST from a checked expression file.
+// The file must be in one of the following formats:
+// - .binarypb
+// - .textproto
+func (c *CompiledExpression) CreateAst(_ Compiler) (*cel.Ast, map[string]any, error) {
+	var expr exprpb.CheckedExpr
+	format := inferFileFormat(c.Path)
+	if format != BinaryProto && format != TextProto {
+		return nil, nil, fmt.Errorf("file extension must be .binarypb or .textproto: found %v", format)
+	}
+	if err := loadProtoFile(c.Path, format, &expr); err != nil {
+		return nil, nil, err
+	}
+	return cel.CheckedExprToAst(&expr), nil, nil
+}
+
+// FileExpression is an InputExpression which loads a CEL expression or policy from a file.
+type FileExpression struct {
+	Path string
+}
+
+// CreateAst creates a CEL AST from a file using the provided compiler:
+// - All policy metadata options as executed using the policy metadata map to extend the
+// environment.
+// - All policy compiler options are passed on to compile the parsed policy.
+//
+// The file must be in one of the following formats:
+// - .cel: CEL string expression
+// - .celpolicy: CEL policy
+func (f *FileExpression) CreateAst(compiler Compiler) (*cel.Ast, map[string]any, error) {
```

Comment:
nit: casing on `Ast` should be `AST` (sorry for the bad precedent from CEL early on with `cel.Ast`)